src/app/clusters/messages-server/messages-server.cpp

```diff
@@ -182,9 +182,51 @@ bool emberAfMessagesClusterPresentMessagesRequestCallback(
     auto & responses      = commandData.responses;
 
     Delegate * delegate = GetDelegate(endpoint);
-    VerifyOrExit(isDelegateNull(delegate, endpoint) != true, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(isDelegateNull(delegate, endpoint) != true, status = Status::NotFound);
+
+    VerifyOrExit(messageId.size() == kMessageIdLength,
+                 ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback invalid message id length");
+                 status = Status::ConstraintError);
+
+    VerifyOrExit(messageText.size() <= kMessageTextLengthMax,
+                 ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback invalid message text length");
+                 status = Status::ConstraintError);
+
+    if (responses.HasValue())
+    {
+        size_t size = 0;
+        err         = responses.Value().ComputeSize(&size);
+        VerifyOrExit(err == CHIP_NO_ERROR,
+                     ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback size check failed");
+                     status = Status::ConstraintError);
+
+        VerifyOrExit(size <= kMessageMaxOptionCount,
+                     ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback too many options");
+                     status = Status::ConstraintError);
+
+        auto iter = responses.Value().begin();
+        while (iter.Next())
+        {
+            auto & response = iter.GetValue();
+
+            VerifyOrExit(response.messageResponseID.HasValue() && response.label.HasValue(),
+                         ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback missing response id or label");
+                         status = Status::InvalidDataType);
+
+            VerifyOrExit(response.messageResponseID.Value() >= kMessageResponseIdMin,
+                         ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback responseID value check failed");
+                         status = Status::ConstraintError);
 
-    delegate->HandlePresentMessagesRequest(messageId, priority, messageControl, startTime, duration, messageText, responses);
+            VerifyOrExit(response.label.Value().size() <= kMessageResponseLabelMaxLength,
+                         ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback label length check failed");
+                         status = Status::ConstraintError);
+        }
+        VerifyOrExit(iter.GetStatus() == CHIP_NO_ERROR,
+                     ChipLogProgress(Zcl, "emberAfMessagesClusterPresentMessagesRequestCallback TLV parsing error");
+                     status = Status::InvalidDataType);
```

Comment:
Malformed encoding should be InvalidAction.

Reply:
addressed